xtask/src/main.rs

```diff
@@ -1,34 +1,125 @@
-use std::str::FromStr;
+use crate::task::Task;
+use anyhow::Result;
+use clap::{Parser, Subcommand};
+use reporter::{BenchmarkReport, ReporterType};
+use std::path::PathBuf;
+use task::{compile::Compile, run::Run};
+use tempfile::{tempdir, TempDir};
+use xshell::{cmd, Shell};
 
-use anyhow::anyhow;
-use metrics::Metrics;
-use xshell::Shell;
+mod reporter;
+mod task;
 
-mod metrics;
-mod traits;
+#[derive(Parser)]
+#[command(author, version, about, long_about = None)]
+struct Parameters {
+    #[command(subcommand)]
+    command: Command,
+}
+
+#[derive(Subcommand, Clone)]
+enum Command {
+    Metrics {
+        #[command(subcommand)]
+        action: Option<Action>,
+        #[arg(value_enum, long, global = true, default_value_t = ReporterType::Sysout)]
+        reporter: ReporterType,
+    },
+}
 
-enum Task {
-    Metrics,
+#[derive(Subcommand, Clone)]
+enum Action {
+    Run {
+        #[arg()]
+        directory: String,
+    },
+    Compile {
+        directory: String,
+    },
 }
 
-impl FromStr for Task {
-    type Err = anyhow::Error;
+fn main() -> anyhow::Result<()> {
+    let params = Parameters::parse();
+    let Command::Metrics { action, reporter } = params.command;
+    run_metrics(action, reporter)?;
+
+    Ok(())
+}
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        match s {
-            "metrics" => Ok(Task::Metrics),
-            _ => Err(anyhow!("Unrecognized task '{s}'")),
+fn run_metrics(action: Option<Action>, reporter: ReporterType) -> Result<()> {
+    let (work_dir, compiler) = prepare()?;
+    //Create tasks
+    let mut tasks: Vec<Box<dyn Task>> = vec![];
+    match &action {
+        Some(Action::Compile { directory }) => {
+            for opt in ["none", "less", "default", "aggressive"] {
+                let task = Compile {
+                    name: directory.clone(),
+                    directory: directory.into(),
+                    optimization: opt.to_string(),
+                    compiler: compiler.clone(),
+                };
+                tasks.push(Box::new(task));
+            }
+        }
+        Some(Action::Run { directory }) => {
+            for opt in ["none", "less", "default", "aggressive"] {
+                let task = Run {
+                    name: directory.into(),
+                    optimization: opt.to_string(),
+                    compiler: compiler.clone(),
+                    location: directory.into(),
+                    parameters: Some("--linker=cc".into()),
+                    work_dir: work_dir.path().into(),
+                };
+                tasks.push(Box::new(task));
+            }
         }
+        None => {
+            //Clone the extra required code
+            println!("Clone Oscat into the benchmarks");
+            let sh = Shell::new()?;
+            let path = work_dir.path();
+            cmd!(&sh, "git clone https://github.com/plc-lang/oscat --depth 1 {path}/oscat").run()?;
```

Comment:
Would it be cleaner to move these lines into the `get_default_tasks` function?

Reply:
Good idea , done